codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -925,5 +920,306 @@ final void incrementStreamableBytesForTree(int numBytes) {
          * Indicates whether or not there are frames in the pending queue.
          */
         abstract boolean hasFrame();
+
+        /**
+         * Returns the parameter from the last call to {@link #markWritabilty(boolean)} (if supported).
+         * {@code false} if not supported.
+         */
+        abstract boolean markWritabilty();
+
+        /**
+         * Save the state of writability (if this method is supported).
+         */
+        abstract void markWritabilty(boolean isWritable);
+    }
+
+    /**
+     * All events which may impact the writability of a {@link Http2Stream} go through this interface.
+     * Implementations can choose how to calculate writability or ignore it all together.
+     */
+    private interface WritabilityMonitor {
+        void pendingBytesChanged(int delta);
+
+        void incrementWindowSize(AbstractState state, int delta) throws Http2Exception;
+
+        void initialWindowSize(int newWindowSize) throws Http2Exception;
+
+        void writePendingBytes() throws Http2Exception;
+
+        void enqueueFrame(AbstractState state, FlowControlled frame) throws Http2Exception;
+
+        boolean isWritable(AbstractState state);
+
+        void stateCancelled(AbstractState state);
+
+        Listener listener();
+
+        void channelWritabilityChange() throws Http2Exception;
+    }
+
+    /**
+     * Abstract class which provides common functionality for {@link WritabilityMonitor} implementations.
+     */
+    private abstract class AbstractWritabilityMonitor implements WritabilityMonitor {
+        protected final void writePendingBytes(Http2StreamVisitor writeAllocatedBytesVisitor) throws Http2Exception {
+            if (ctx == null) {
```

Comment:
Good question. I think it is possible that `initialWindowSize` could be called before the channel context is set through `DefaultHttp2ConnectionDecoder.localSettings`.